### ISSUE TITLE
[Bug?]: Yarns patching of the fs module subtly breaks in node >20.5.1

### ISSUE THREAD

**Issue by danieltroger** [yarnpkg/berry#5951]
### Self-service

- [ ] I'd be willing to implement a fix

### Describe the bug

On node versions above 20.5.1, swc-node stops working with the following cryptic error:

```
node:internal/process/esm_loader:40
      internalBinding('errors').triggerUncaughtException(
                                ^
TypeError [ERR_INVALID_RETURN_PROPERTY_VALUE]: Expected array buffer, or typed array to be returned for the "source" from the "transformSource" function but got undefined.
    at new NodeError (node:internal/errors:406:5)
    at assertBufferSource (node:internal/modules/esm/translators:84:9)
    at stringify (node:internal/modules/esm/translators:94:3)
    at createCJSModuleWrap (node:internal/modules/esm/translators:219:12)
    at ModuleLoader.commonjsStrategy (node:internal/modules/esm/translators:297:10) {
  code: 'ERR_INVALID_RETURN_PROPERTY_VALUE'
}

Node.js v20.9.0
```

This is due to the node js worker which deals with something related to custom esm modules has a reference to the unpatched `readFileSync` function from the `fs` module, see https://github.com/swc-project/swc-node/issues/736#issuecomment-1800239591

### To reproduce

Follow the steps outlined in this issue report: https://github.com/swc-project/swc-node/issues/736#issue-1962957988

### Environment

```shell
System:
  OS: macOS 14.0 (23A344)
  CPU: Apple M1 Max
Binaries:
  Node: v20.9.0
  Yarn: 4.0.1-git.20231104.hash-afacd23
```

**Comment by 0618** [comment]
I have the same issue.

Environment:

```
  System:
    OS: macOS 13.6
    CPU: (12) arm64 Apple M2 Pro
  Binaries:
    Node: 20.9.0 - /private/var/folders/0m/fgfvh9sj5d9fvwzhvd8khpf80000gr/T/xfs-4f6f8be2/node
    Yarn: 4.0.1 - /private/var/folders/0m/fgfvh9sj5d9fvwzhvd8khpf80000gr/T/xfs-4f6f8be2/yarn
```

**Comment by vhpoet** [comment]
Same here

Environment:
```
  System:
    OS: macOS 14.0
    CPU: Apple M1 Max
  Binaries:
    Node: 21.1.0
    Yarn: 4.0.1
```

**Comment by arcanis** [comment]
This is a Node bug which they don't see as a bug 🙁 (https://github.com/nodejs/node/pull/39513)

**Comment by danieltroger** [comment]
@arcanis thanks for the link. Holy guacamole, what a situation. I have two questions:

1. Why does this first start breaking swc-node above node 20.5.0? From the issue you linked it appears to be quite an old issue
2. Is there no way to work around this via loaders or some other node feature?

**Comment by arcanis** [comment]
I'm not sure unfortunately. I've been pointed out to another issue on the Node.js Slack channels which, together with the first one, may be why this problem just appeared: https://github.com/nodejs/node/issues/50435.

**Comment by lizthegrey** [comment]
> @arcanis thanks for the link. Holy guacamole, what a situation. I have two questions:
> 
> 1. Why does this first start breaking swc-node above node 20.5.0? From the issue you linked it appears to be quite an old issue
> 2. Is there no way to work around this via loaders or some other node feature? Is there a way of using swc-node that doesn't run into this issue?

This will be fixed in Node 20.11.0, per https://github.com/nodejs/node/issues/50435#issuecomment-1855256713
